### ISSUE TITLE
Padding should use 2 cache lines to pad

### ISSUE THREAD

**Issue by franz1981** [LMAX-Exchange/disruptor#231]
HI!!!

given that prefetchers use 2 cache lines, `Sequence` (or similar padded classes) shouldn't be instead:
```
    static class LhsPadding
    {
        protected long p1, p2, p3, p4, p5, p6, p7;
        protected long p8, p9, p10, p11, p12, p13, p14, p15;
    }

    static class Value extends LhsPadding
    {
        protected volatile long value;
    }

    static class RhsPadding extends Value
    {
        protected long p1, p2, p3, p4, p5, p6, p7;
        protected long p8, p9, p10, p11, p12, p13, p14, p15;
    }
```

To allow with both compressed oop or not:
```
RhsPadding object internals:
 OFFSET  SIZE   TYPE DESCRIPTION                               VALUE
      0    12        (object header)                           N/A
     12     4        (alignment/padding gap)                  
     16     8   long LhsPadding.p1                             N/A
     24     8   long LhsPadding.p2                             N/A
     32     8   long LhsPadding.p3                             N/A
     40     8   long LhsPadding.p4                             N/A
     48     8   long LhsPadding.p5                             N/A
     56     8   long LhsPadding.p6                             N/A
     64     8   long LhsPadding.p7                             N/A
     72     8   long LhsPadding.p8                             N/A
     80     8   long LhsPadding.p9                             N/A
     88     8   long LhsPadding.p10                            N/A
     96     8   long LhsPadding.p11                            N/A
    104     8   long LhsPadding.p12                            N/A
    112     8   long LhsPadding.p13                            N/A
    120     8   long LhsPadding.p14                            N/A
    128     8   long LhsPadding.p15                            N/A
    136     8   long Value.value                               N/A
    144     8   long RhsPadding.p1                             N/A
    152     8   long RhsPadding.p2                             N/A
    160     8   long RhsPadding.p3                             N/A
    168     8   long RhsPadding.p4                             N/A
    176     8   long RhsPadding.p5                             N/A
    184     8   long RhsPadding.p6                             N/A
    192     8   long RhsPadding.p7                             N/A
    200     8   long RhsPadding.p8                             N/A
    208     8   long RhsPadding.p9                             N/A
    216     8   long RhsPadding.p10                            N/A
    224     8   long RhsPadding.p11                            N/A
    232     8   long RhsPadding.p12                            N/A
    240     8   long RhsPadding.p13                            N/A
    248     8   long RhsPadding.p14                            N/A
    256     8   long RhsPadding.p15                            N/A
Instance size: 264 bytes
```
On `LhsPadding` at 8 bytes boundaries in the worst case a cache line due to an access to the last bytes of the object header would start at offset 8 ie `Value.value` should start at 136 to avoid being 
prefetched with it.
The same reasoning should apply on `RhsPadding` as well.
In addition, after an answer on Twitter with Aleksey Shipilëv it seems that using  bytes to pad (as JMH does) would be better:

> Also, don't pad with longs: can you see those gaps that could theoretically be taken by the int/compressed-oops fields in your example? JMH, for instance, pads with bytes.

TBH in this case I would maintain `long` , because with uncompressed oops the `(alignment/padding gap)` will just disappear, while leaving the first critical cache line start as it is ie 8 bytes offset, but maybe I've missed the true reason behind his answer.

**Comment by franz1981** [comment]
This is issue sounds more like a question TBH: if padding with 64 bytes will prove to be enough from the performance pov I think that won't be necessary to do differently :+1:

**Comment by Palmr** [comment]
It's been a while since this ticket was opened, but I was keen to look in to this.

In terms of the prefetcher on modern Intel CPU's fetching two cache lines, this is true. I attempted to benchmark this locally on my `Intel(R) Core(TM) i5-6600K CPU @ 3.50GHz` (`Adjacent Cache Line Prefetch` enabled in bios) with the following JMH: https://gist.github.com/Palmr/0729b4c8b8f422487b3fd47d4cfe690b

My results were as follows:

```
Benchmark                                                      Mode  Cnt     Score    Error   Units
FalseSharingCacheLinePaddingBenchmark.baseline                thrpt   25  1631.217 ± 25.471  ops/us
FalseSharingCacheLinePaddingBenchmark.baseline:reader         thrpt   25   199.340 ± 14.455  ops/us
FalseSharingCacheLinePaddingBenchmark.baseline:writer         thrpt   25  1431.877 ± 29.613  ops/us
FalseSharingCacheLinePaddingBenchmark.doubleCacheLine         thrpt   25  2387.148 ± 21.451  ops/us
FalseSharingCacheLinePaddingBenchmark.doubleCacheLine:reader  thrpt   25   837.000 ± 12.515  ops/us
FalseSharingCacheLinePaddingBenchmark.doubleCacheLine:writer  thrpt   25  1550.149 ± 17.373  ops/us
FalseSharingCacheLinePaddingBenchmark.singleCacheLine         thrpt   25  2337.514 ± 46.727  ops/us
FalseSharingCacheLinePaddingBenchmark.singleCacheLine:reader  thrpt   25   830.081 ± 29.697  ops/us
FalseSharingCacheLinePaddingBenchmark.singleCacheLine:writer  thrpt   25  1507.433 ± 25.438  ops/us
```

This made me pause and confer with some other people, as the single and double cache line padding shows no notable difference.
In the benchmark I had two threads reading and writing to two separate values some padding apart. The adjacent cache line that is fetched gets marked invalid/dirty when changed by the other thread. But since it's not otherwise touched there's no cost to this.

Given the `Sequence` has left & right padding about the actual field, if you had two `Sequence` instances next to each other in memory there would be a 2-cache line padding between them I think? So we're safe there anyway.

For other values which may end up in the adjacent cache line to a Sequence, if it's not touched by both threads it's not an issue. It would be like the above benchmark where the adjacent cache line only will get marked invalid/dirty and since the other thread isn't looking at those values it has no need to request a read from them and pay a cost.
The cache line it is modifying, when it gets marked invalid/dirty by another thread, will just be requested a read via the cache protocol and updated. As it should.

I think if we were going to consider making any change here we'd need a benchmark proving there is a clear impact from Adjacent Cache Line Prefetch on `Sequence` which is fixed by padding more or disabling Adjacent Cache Line Prefetch.

**Comment by Palmr** [comment]
I was interested in the mention of this

> Also, don't pad with longs: can you see those gaps that could theoretically be taken by the int/compressed-oops fields in your example? JMH, for instance, pads with bytes.

So I ended up reading https://shipilev.net/jvm/objects-inside-out/#_superhierarchy_gaps_in_java_15 and https://bugs.openjdk.java.net/browse/JDK-8237767

From what I have read they have changed the field layout code in HotSpot 15. 
Before the object hierarchy trick always held true, fields in a super class would always be laid out in memory before fields in the extending class. Even if this meant an alignment/padding gap.
With HotSpot 15 it has changed and fields can be re-ordered with super-class fields.

Relying on the field layout based on class hierarchy as we do is risky. But the only other alternatives are the `@Contended` annotation, which is not made for public use, or using an array which comes with significant performance impact.

Looking in to it further I see that the change to re-order fields so that they can be mixed in to super-class fields is there for efficient packing purposes. If there no object layout size advantage to be gained, this doesn't happen.

**This is the case for the current padding structure in `Sequence`, there will be no re-ordering of fields by HotSpot, even with JDK15, as all fields are the same type and being `long` types they don't fit in to any alignment/padding gaps.**

`Sequence` is padded with `long` fields, and the value field we're protecting from false-sharing is also a `long`. This means the packing leaves no gaps and there's no reason for the change in layout method to move the `value` field from where it would have been prior to HotSpot 15.

As an example using OpenJDK 11 and Azul Zulu 15 (both distributions of HotSpot):

## 11
```
$ /usr/lib/jvm/java-11-openjdk/bin/java -Djol.tryWithSudo=true -jar jol-cli.jar internals -cp build/classes/java/main com.lmax.disruptor.Sequence
# Running 64-bit HotSpot VM.
# Using compressed oop with 3-bit shift.
# Using compressed klass with 0x0000000800000000 base address and 0-bit shift.
# Objects are 8 bytes aligned.
# Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
# Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

Instantiated the sample instance via default constructor.

com.lmax.disruptor.Sequence object internals:
 OFFSET  SIZE   TYPE DESCRIPTION                               VALUE
      0     4        (object header)                           05 00 00 00 (00000101 00000000 00000000 00000000) (5)
      4     4        (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4        (object header)                           40 40 0d 00 (01000000 01000000 00001101 00000000) (868416)
     12     4        (alignment/padding gap)                  
     16     8   long LhsPadding.p1                             0
     24     8   long LhsPadding.p2                             0
     32     8   long LhsPadding.p3                             0
     40     8   long LhsPadding.p4                             0
     48     8   long LhsPadding.p5                             0
     56     8   long LhsPadding.p6                             0
     64     8   long LhsPadding.p7                             0
     72     8   long Value.value                               -1
     80     8   long RhsPadding.p9                             0
     88     8   long RhsPadding.p10                            0
     96     8   long RhsPadding.p11                            0
    104     8   long RhsPadding.p12                            0
    112     8   long RhsPadding.p13                            0
    120     8   long RhsPadding.p14                            0
    128     8   long RhsPadding.p15                            0
Instance size: 136 bytes
Space losses: 4 bytes internal + 0 bytes external = 4 bytes total
```

## 15
```
$ ~/.jdks/azul-15.0.1/bin/java -Djol.tryWithSudo=true -jar jol-cli.jar internals -cp build/classes/java/main com.lmax.disruptor.Sequence
# WARNING: Unable to attach Serviceability Agent. sun.jvm.hotspot.memory.Universe.getNarrowOopBase()
# Running 64-bit HotSpot VM.
# Using compressed oop with 3-bit shift.
# Using compressed klass with 3-bit shift.
# WARNING | Compressed references base/shifts are guessed by the experiment!
# WARNING | Therefore, computed addresses are just guesses, and ARE NOT RELIABLE.
# WARNING | Make sure to attach Serviceability Agent to get the reliable addresses.
# Objects are 8 bytes aligned.
# Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
# Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

Instantiated the sample instance via default constructor.

com.lmax.disruptor.Sequence object internals:
 OFFSET  SIZE   TYPE DESCRIPTION                               VALUE
      0     4        (object header)                           01 00 00 00 (00000001 00000000 00000000 00000000) (1)
      4     4        (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4        (object header)                           08 c1 17 00 (00001000 11000001 00010111 00000000) (1556744)
     12     4        (alignment/padding gap)                  
     16     8   long LhsPadding.p1                             0
     24     8   long LhsPadding.p2                             0
     32     8   long LhsPadding.p3                             0
     40     8   long LhsPadding.p4                             0
     48     8   long LhsPadding.p5                             0
     56     8   long LhsPadding.p6                             0
     64     8   long LhsPadding.p7                             0
     72     8   long Value.value                               -1
     80     8   long RhsPadding.p9                             0
     88     8   long RhsPadding.p10                            0
     96     8   long RhsPadding.p11                            0
    104     8   long RhsPadding.p12                            0
    112     8   long RhsPadding.p13                            0
    120     8   long RhsPadding.p14                            0
    128     8   long RhsPadding.p15                            0
Instance size: 136 bytes
Space losses: 4 bytes internal + 0 bytes external = 4 bytes total
```

If the `value` in `Sequence` was a smaller data type, i.e. `int` instead of `long`, we might see a bad layout for us from HotSpot 15 as the layout method can find a place for a smaller data type in the alignment/padding gap at offset 12

## 11
```
$ /usr/lib/jvm/java-11-openjdk/bin/java -Djol.tryWithSudo=true -jar jol-cli.jar internals -cp build/classes/java/main com.lmax.disruptor.Sequence
# Running 64-bit HotSpot VM.
# Using compressed oop with 3-bit shift.
# Using compressed klass with 0x0000000800000000 base address and 0-bit shift.
# Objects are 8 bytes aligned.
# Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
# Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

Instantiated the sample instance via default constructor.

com.lmax.disruptor.Sequence object internals:
 OFFSET  SIZE   TYPE DESCRIPTION                               VALUE
      0     4        (object header)                           05 00 00 00 (00000101 00000000 00000000 00000000) (5)
      4     4        (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4        (object header)                           40 40 0d 00 (01000000 01000000 00001101 00000000) (868416)
     12     4        (alignment/padding gap)                  
     16     8   long LhsPadding.p1                             0
     24     8   long LhsPadding.p2                             0
     32     8   long LhsPadding.p3                             0
     40     8   long LhsPadding.p4                             0
     48     8   long LhsPadding.p5                             0
     56     8   long LhsPadding.p6                             0
     64     8   long LhsPadding.p7                             0
     72     4    int Value.value                               -1
     76     4        (alignment/padding gap)                  
     80     8   long RhsPadding.p9                             0
     88     8   long RhsPadding.p10                            0
     96     8   long RhsPadding.p11                            0
    104     8   long RhsPadding.p12                            0
    112     8   long RhsPadding.p13                            0
    120     8   long RhsPadding.p14                            0
    128     8   long RhsPadding.p15                            0
Instance size: 136 bytes
Space losses: 8 bytes internal + 0 bytes external = 8 bytes total
```

## 15
```
$ ~/.jdks/azul-15.0.1/bin/java -Djol.tryWithSudo=true -jar jol-cli.jar internals -cp build/classes/java/main com.lmax.disruptor.Sequence
# WARNING: Unable to attach Serviceability Agent. sun.jvm.hotspot.memory.Universe.getNarrowOopBase()
# Running 64-bit HotSpot VM.
# Using compressed oop with 3-bit shift.
# Using compressed klass with 3-bit shift.
# WARNING | Compressed references base/shifts are guessed by the experiment!
# WARNING | Therefore, computed addresses are just guesses, and ARE NOT RELIABLE.
# WARNING | Make sure to attach Serviceability Agent to get the reliable addresses.
# Objects are 8 bytes aligned.
# Field sizes by type: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]
# Array element sizes: 4, 1, 1, 2, 2, 4, 4, 8, 8 [bytes]

Instantiated the sample instance via default constructor.

com.lmax.disruptor.Sequence object internals:
 OFFSET  SIZE   TYPE DESCRIPTION                               VALUE
      0     4        (object header)                           01 00 00 00 (00000001 00000000 00000000 00000000) (1)
      4     4        (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4        (object header)                           08 c1 17 00 (00001000 11000001 00010111 00000000) (1556744)
     12     4    int Value.value                               -1
     16     8   long LhsPadding.p1                             0
     24     8   long LhsPadding.p2                             0
     32     8   long LhsPadding.p3                             0
     40     8   long LhsPadding.p4                             0
     48     8   long LhsPadding.p5                             0
     56     8   long LhsPadding.p6                             0
     64     8   long LhsPadding.p7                             0
     72     8   long RhsPadding.p9                             0
     80     8   long RhsPadding.p10                            0
     88     8   long RhsPadding.p11                            0
     96     8   long RhsPadding.p12                            0
    104     8   long RhsPadding.p13                            0
    112     8   long RhsPadding.p14                            0
    120     8   long RhsPadding.p15                            0
Instance size: 128 bytes
Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
```

**Comment by Palmr** [comment]
I think I have satisfied myself that this issue is a no-op for us.

 - Benchmarks appear to show no improvement with padding twice the size of cache lines to avoid issues with the Intel prefetcher fetching adjacent lines
 - `jol` has shown me that field re-ordering is not an issue for `Sequence` with HotSpot 15

Therefore I think I can close this issues now.
These were good questions though and I enjoyed digging in to them, if there are any more thoughts / suggestions let us know.
